File: src/test/java/com/cxxcxx/zinecraft/api/accessory/CollectibleTooltipWrappingTest.java
package com.cxxcxx.zinecraft.api.accessory;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class CollectibleTooltipWrappingTest {
  @Test
  void pairsDifferentLocalizationLengthsWithoutChangingVisibleText() {
    String zhCn = "短中文原文";
    String enUs = "A deliberately longer English localization that needs several independently rendered tooltip lines.";
    List<LocalizedTooltipLine> lines = CollectibleTooltips.wrapLocalizedTooltip(zhCn, enUs, 24, 42);

    assertTrue(lines.size() > 1);
    assertEquals(zhCn, lines.stream().map(LocalizedTooltipLine::getZhCn).reduce("", String::concat).replace("\u200B", ""));
    assertEquals(enUs, lines.stream().map(LocalizedTooltipLine::getEnUs).reduce("", String::concat).replace("\u200B", ""));
    assertTrue(lines.getFirst().getZhCn().codePointCount(0, lines.getFirst().getZhCn().length()) <= 24);
    assertTrue(lines.getFirst().getEnUs().codePointCount(0, lines.getFirst().getEnUs().length()) <= 24);
    assertTrue(lines.stream().skip(1).allMatch(line -> line.getZhCn().length() <= 42 && line.getEnUs().length() <= 42));
  }
}

File: src/test/kotlin/com/cxxcxx/zinecraft/api/accessory/CollectibleTooltipWrappingTest.kt
package com.cxxcxx.zinecraft.api.accessory

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class CollectibleTooltipWrappingTest {
  @Test
  fun `pairs different localization lengths without changing visible text`() {
    val zhCn = "短中文原文"
    val enUs = "A deliberately longer English localization that needs several independently rendered tooltip lines."
    val lines = wrapLocalizedTooltip(zhCn, enUs, firstLineCharacters = 24, continuationCharacters = 42)

    assertTrue(lines.size > 1)
    assertEquals(zhCn, lines.joinToString("") { it.zhCn }.replace("\u200B", ""))
    assertEquals(enUs, lines.joinToString("") { it.enUs }.replace("\u200B", ""))
    assertTrue(lines.first().zhCn.codePointCount(0, lines.first().zhCn.length) <= 24)
    assertTrue(lines.first().enUs.codePointCount(0, lines.first().enUs.length) <= 24)
    assertTrue(lines.drop(1).all { it.zhCn.length <= 42 && it.enUs.length <= 42 })
  }
}
